In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
DATA_PATH_NUM = '/content/drive/My Drive/german_credit_project/data/raw/german.data-numeric'
df_num = pd.read_csv(DATA_PATH_NUM, sep='\s+', header = None)
df_num.head()

<>:3: SyntaxWarning: invalid escape sequence '\s'
<>:3: SyntaxWarning: invalid escape sequence '\s'
/tmp/ipykernel_8285/3457253743.py:3: SyntaxWarning: invalid escape sequence '\s'
  df_num = pd.read_csv(DATA_PATH_NUM, sep='\s+', header = None)


,0,1,2,3,4,5,6,7,8,9,...,15,16,17,18,19,20,21,22,23,24
0,1,6,4,12,5,5,3,4,1,67,...,0,0,1,0,0,1,0,0,1,1
1,2,48,2,60,1,3,2,2,1,22,...,0,0,1,0,0,1,0,0,1,2
2,4,12,4,21,1,4,3,3,1,49,...,0,0,1,0,0,1,0,1,0,1
3,1,42,2,79,1,4,3,4,2,45,...,0,0,0,0,0,0,0,0,1,1
4,1,24,3,49,1,3,3,4,4,53,...,1,0,1,0,0,0,0,0,1,2


In [ ]:
import pandas as pd
DATA_PATH_ORG = '/content/drive/My Drive/german_credit_project/data/raw/german.data'
df_org = pd.read_csv(DATA_PATH_ORG,sep='\s+', header=None)
df_org.head()

<>:3: SyntaxWarning: invalid escape sequence '\s'
<>:3: SyntaxWarning: invalid escape sequence '\s'
/tmp/ipykernel_8285/3362526608.py:3: SyntaxWarning: invalid escape sequence '\s'
  df_org = pd.read_csv(DATA_PATH_ORG,sep='\s+', header=None)


,0,1,2,3,4,5,6,7,8,9,...,11,12,13,14,15,16,17,18,19,20
0,A11,6,A34,A43,1169,A65,A75,4,A93,A101,...,A121,67,A143,A152,2,A173,1,A192,A201,1
1,A12,48,A32,A43,5951,A61,A73,2,A92,A101,...,A121,22,A143,A152,1,A173,1,A191,A201,2
2,A14,12,A34,A46,2096,A61,A74,2,A93,A101,...,A121,49,A143,A152,1,A172,2,A191,A201,1
3,A11,42,A32,A42,7882,A61,A74,2,A93,A103,...,A122,45,A143,A153,1,A173,2,A191,A201,1
4,A11,24,A33,A40,4870,A61,A73,3,A93,A101,...,A124,53,A143,A153,2,A173,2,A191,A201,2


# 1. Data Preprocessing

In [ ]:


# Convert target variable to binary format (0=low risk, 1=high risk)
df_org['credit_risk'] = df_org[20].map({1:0, 2:1})

# drop the original column
df_org.drop(20, axis=1, inplace=True)


# Display the updated dataframe
df_org.head()

,0,1,2,3,4,5,6,7,8,9,...,11,12,13,14,15,16,17,18,19,credit_risk
0,A11,6,A34,A43,1169,A65,A75,4,A93,A101,...,A121,67,A143,A152,2,A173,1,A192,A201,0
1,A12,48,A32,A43,5951,A61,A73,2,A92,A101,...,A121,22,A143,A152,1,A173,1,A191,A201,1
2,A14,12,A34,A46,2096,A61,A74,2,A93,A101,...,A121,49,A143,A152,1,A172,2,A191,A201,0
3,A11,42,A32,A42,7882,A61,A74,2,A93,A103,...,A122,45,A143,A153,1,A173,2,A191,A201,0
4,A11,24,A33,A40,4870,A61,A73,3,A93,A101,...,A124,53,A143,A153,2,A173,2,A191,A201,1


## 1.1 Column Renaming

In [ ]:

# Renaming the columns
df_org = df_org.rename(columns = {
  0:  "Status of existing checking account",
  1: "Duration in month",
  2: "Credit History",
  3: "Purpose",
  4: "Credit amount",
  5: "Savings account/bonds",
  6: "Present employment since",
  7: "Installment rate in percentage of disposable income",
  8: "Personal status and sex",
  9: "Other debtors /guarantors",
  10: "Present residence since",
  11: "Property",
  12: "Age in years",
  13: "Other installment plans",
  14: "Housing",
  15: "Number of existing credits at this bank",
  16: "Job",
  17: "Number of people being liable to provide maintenance for",
  18: "Telephone",
  19: "Foreign worker"


})

print(df_org.columns)


df_new = df_org.copy()

Index(['Status of existing checking account', 'Duration in month',
       'Credit History', 'Purpose', 'Credit amount', 'Savings account/bonds',
       'Present employment since',
       'Installment rate in percentage of disposable income',
       'Personal status and sex', 'Other debtors /guarantors',
       'Present residence since', 'Property', 'Age in years',
       'Other installment plans', 'Housing',
       'Number of existing credits at this bank', 'Job',
       'Number of people being liable to provide maintenance for', 'Telephone',
       'Foreign worker', 'credit_risk'],
      dtype='object')


## 1.2 One hot encoding the categorical values

In [ ]:
# One hot encoding
import pandas as pd
from sklearn.preprocessing import OneHotEncoder

cat_cols = df_new.select_dtypes(include=['object']).columns

encoder = OneHotEncoder(sparse_output=False)

encoded_data = encoder.fit_transform(df_new[cat_cols])

encoded_df = pd.DataFrame(
    encoded_data,
    columns = encoder.get_feature_names_out(cat_cols)
)

final_df = pd.concat(
    [df_new.drop(columns=cat_cols), encoded_df],
    axis=1
)

print("\nOne-Hot Encoded Data:")
final_df.head()





One-Hot Encoded Data:


,Duration in month,Credit amount,Installment rate in percentage of disposable income,Present residence since,Age in years,Number of existing credits at this bank,Number of people being liable to provide maintenance for,credit_risk,Status of existing checking account_A11,Status of existing checking account_A12,...,Housing_A152,Housing_A153,Job_A171,Job_A172,Job_A173,Job_A174,Telephone_A191,Telephone_A192,Foreign worker_A201,Foreign worker_A202
0,6,1169,4,4,67,2,1,0,1.0,0.0,...,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
1,48,5951,2,2,22,1,1,1,0.0,1.0,...,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
2,12,2096,2,3,49,1,2,0,0.0,0.0,...,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0
3,42,7882,2,4,45,1,2,0,1.0,0.0,...,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
4,24,4870,3,4,53,2,2,1,1.0,0.0,...,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0


In [ ]:
print(final_df.shape)

(1000, 62)


In [ ]:
print(final_df.columns.tolist())

['Duration in month', 'Credit amount', 'Installment rate in percentage of disposable income', 'Present residence since', 'Age in years', 'Number of existing credits at this bank', 'Number of people being liable to provide maintenance for', 'credit_risk', 'Status of existing checking account_A11', 'Status of existing checking account_A12', 'Status of existing checking account_A13', 'Status of existing checking account_A14', 'Credit History_A30', 'Credit History_A31', 'Credit History_A32', 'Credit History_A33', 'Credit History_A34', 'Purpose_A40', 'Purpose_A41', 'Purpose_A410', 'Purpose_A42', 'Purpose_A43', 'Purpose_A44', 'Purpose_A45', 'Purpose_A46', 'Purpose_A48', 'Purpose_A49', 'Savings account/bonds_A61', 'Savings account/bonds_A62', 'Savings account/bonds_A63', 'Savings account/bonds_A64', 'Savings account/bonds_A65', 'Present employment since_A71', 'Present employment since_A72', 'Present employment since_A73', 'Present employment since_A74', 'Present employment since_A75', 'Person

In [ ]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 62 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Duration in month                                         1000 non-null   int64  
 1   Credit amount                                             1000 non-null   int64  
 2   Installment rate in percentage of disposable income       1000 non-null   int64  
 3   Present residence since                                   1000 non-null   int64  
 4   Age in years                                              1000 non-null   int64  
 5   Number of existing credits at this bank                   1000 non-null   int64  
 6   Number of people being liable to provide maintenance for  1000 non-null   int64  
 7   credit_risk                                               1000 non-null   int64  
 8   Status of existing 

In [ ]:
rename_dict = {

# Status of existing checking account
'Status of existing checking account_A11': 'checking_balance_negative',
'Status of existing checking account_A12': 'checking_balance_0_200DM',
'Status of existing checking account_A13': 'checking_balance_ge_200DM',
'Status of existing checking account_A14': 'no_checking_account',


# Credit History
'Credit History_A30': 'credit_all_paid',
'Credit History_A31': 'credit_all_paid_bank',
'Credit History_A32': 'credit_paid_current',
'Credit History_A33': 'credit_past_delay',
'Credit History_A34': 'credit_critical_account',

# Purpose
'Purpose_A40': 'purpose_car(new)',
'Purpose_A41': 'purpose_car(old)',
'Purpose_A42': 'purpose_furniture',
'Purpose_A43': 'purpose_radio/tv',
'Purpose_A44': 'purpose_domestic_appliances',
'Purpose_A45': 'purpose_repairs',
'Purpose_A46': 'purpose_education',
'Purpose_A47': 'purpose_vacation',
'Purpose_A48': 'purpose_retraining',
'Purpose_A49': 'purpose_business',
'Purpose_A410': 'purpose_others',


# Savings account/bonds
'Savings account/bonds_A61': 'savings_balance_lt_100DM',
'Savings account/bonds_A62': 'savings_balance_100_to_500DM',
'Savings account/bonds_A63': 'savings_balance_500_to_1000DM',
'Savings account/bonds_A64': 'savings_balance_ge_1000DM',
'Savings account/bonds_A65': 'no_or_unknown_savings',

# Present employment since
'Present employment since_A71': 'employment_duration_unemployed',
'Present employment since_A72': 'employment_duration_lt_1yr',
'Present employment since_A73': 'employment_duration_1_to_4yrs',
'Present employment since_A74': 'employment_duration_4_to_7yrs',
'Present employment since_A75': 'employment_duration_gte_7yrs',

# Personal status and sex
'Personal status and sex_A91': 'male_divorced_separated',
'Personal status and sex_A92': 'female_divorced_separated_married',
'Personal status and sex_A93': 'male_single',
'Personal status and sex_A94': 'male_married_widowed',
'Personal status and sex_A95': 'female_single',

# Other debtors/guarantors
'Other debtors /guarantors_A101': 'guarantors_none',
'Other debtors /guarantors_A102': 'guarantors_co-applicant',
'Other debtors /guarantors_A103': 'guarantors_guarantor',

# Property
'Property_A121': 'property_real_estate',
'Property_A122': 'property_building_savings_life_insurance',
'Property_A123': 'property_car_or_other',
'Property_A124': 'property_unknown_or_none',

 # Other installment plans
 'Other installment plans_A141': 'installment_plan_bank',
 'Other installment plans_A142': 'installment_plan_store',
 'Other installment plans_A143': 'no_installment_plan',

 # Housing
 'Housing_A151': 'housing_rent',
 'Housing_A152': 'housing_own',
 'Housing_A153': 'housing_free',

 # Job
 'Job_A171': 'job_unemployed_unskilled_nonresident',
 'Job_A172': 'job_unskilled_resident',
 'Job_A173': 'job_skilled_employee',
 'Job_A174': 'job_management_self_employed_highly_qualified',

 # Telephone
 'Telephone_A191': 'telephone_none',
 'Telephone_A192': 'telephone_registered',

 # Foreign worker
 'Foreign worker_A201': 'foreign_worker_yes',
 'Foreign worker_A202': 'foreign_worker_no'

}

df_readable = final_df.rename(columns=rename_dict)
df_readable.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 62 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Duration in month                                         1000 non-null   int64  
 1   Credit amount                                             1000 non-null   int64  
 2   Installment rate in percentage of disposable income       1000 non-null   int64  
 3   Present residence since                                   1000 non-null   int64  
 4   Age in years                                              1000 non-null   int64  
 5   Number of existing credits at this bank                   1000 non-null   int64  
 6   Number of people being liable to provide maintenance for  1000 non-null   int64  
 7   credit_risk                                               1000 non-null   int64  
 8   checking_balance_ne

## Separate features and target


In [ ]:
df_readable.columns

Index(['Duration in month', 'Credit amount',
       'Installment rate in percentage of disposable income',
       'Present residence since', 'Age in years',
       'Number of existing credits at this bank',
       'Number of people being liable to provide maintenance for',
       'credit_risk', 'checking_balance_negative', 'checking_balance_0_200DM',
       'checking_balance_ge_200DM', 'no_checking_account', 'credit_all_paid',
       'credit_all_paid_bank', 'credit_paid_current', 'credit_past_delay',
       'credit_critical_account', 'purpose_car(new)', 'purpose_car(old)',
       'purpose_others', 'purpose_furniture', 'purpose_radio/tv',
       'purpose_domestic_appliances', 'purpose_repairs', 'purpose_education',
       'purpose_retraining', 'purpose_business', 'savings_balance_lt_100DM',
       'savings_balance_100_to_500DM', 'savings_balance_500_to_1000DM',
       'savings_balance_ge_1000DM', 'no_or_unknown_savings',
       'employment_duration_unemployed', 'employment_duration_lt_1

In [ ]:
X = df_readable.drop('credit_risk', axis=1)
y=df_readable['credit_risk']

In [ ]:
# Check for class imbalance
y.value_counts()

,count
credit_risk,
0,700
1,300


In [ ]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 1000 entries, 0 to 999
Series name: credit_risk
Non-Null Count  Dtype
--------------  -----
1000 non-null   int64
dtypes: int64(1)
memory usage: 7.9 KB


In [ ]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 61 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Duration in month                                         1000 non-null   int64  
 1   Credit amount                                             1000 non-null   int64  
 2   Installment rate in percentage of disposable income       1000 non-null   int64  
 3   Present residence since                                   1000 non-null   int64  
 4   Age in years                                              1000 non-null   int64  
 5   Number of existing credits at this bank                   1000 non-null   int64  
 6   Number of people being liable to provide maintenance for  1000 non-null   int64  
 7   checking_balance_negative                                 1000 non-null   float64
 8   checking_balance_0_

## Train-test split

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

## Model Development

6.1 Experiment 1: Baseline Models (Original Dataset)

In [ ]:
# Linear model
from sklearn.linear_model import LogisticRegression
# Tree models
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
# K-Nearest Neighbors
from sklearn.neighbors import KNeighborsClassifier
# Support Vector Machine
from sklearn.svm import SVC
#Naive Bayes
from sklearn.naive_bayes import GaussianNB

# Import metrics
from sklearn.metrics import recall_score, accuracy_score, precision_score, f1_score


algo_list = {
    "log_reg": LogisticRegression(random_state=42),
    "decision_tree": DecisionTreeClassifier(random_state=42),
    "random_forest": RandomForestClassifier(random_state=42),
    "svm_model": SVC(random_state=42),
    "knn": KNeighborsClassifier(),
    "naive_bayes": GaussianNB()
}

results = []

for name, model in algo_list.items():
  model.fit(X_train, y_train)
  pred_val = model.predict(X_test)



  accuracy = accuracy_score(y_test, pred_val)
  recall = recall_score(y_test, pred_val)
  precision = precision_score(y_test, pred_val)
  f1 = f1_score(y_test, pred_val)

  results.append({
      "Model": name,
      "Accuracy": accuracy,
      "Precision": precision,
      "Recall": recall,
      "F1 score": f1
  })







/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [ ]:
results_df = pd.DataFrame(results)
results_df.set_index("Model", inplace=True)
results_df

,Accuracy,Precision,Recall,F1 score
Model,,,,
log_reg,0.785,0.688889,0.516667,0.590476
decision_tree,0.630,0.379310,0.366667,0.372881
random_forest,0.740,0.605263,0.383333,0.469388
svm_model,0.710,0.583333,0.116667,0.194444
knn,0.680,0.447368,0.283333,0.346939
naive_bayes,0.720,0.527027,0.650000,0.582090


### Insight

Logistic Regression performed better than the other baseline models based on accuracy and F1-score. However, since credit risk prediction involves different consequences for incorrect predictions, other metrics should also be considered. Naive Bayes achieved higher recall, meaning it was able to identify more high-risk applicants, while logistic Regression had higher precision, meaning its high-risk predictions were more reliable. Therefore, the preferred model depends on whether the priority is reducing financial risk or avoiding unnecessary rejection of applicants.  

### 6.2 Experiment - 2 (handling class imbalance with weights)


In [ ]:
algo_list = {
    "log_reg": LogisticRegression(random_state=42, class_weight="balanced", max_iter=6200),
    "decision_tree": DecisionTreeClassifier(random_state=42, class_weight="balanced"),
    "random_forest": RandomForestClassifier(random_state=42, class_weight="balanced"),
    "svm_model": SVC(random_state=42, class_weight="balanced")

}

results = []

for name, model in algo_list.items():
  model.fit(X_train, y_train)
  pred_val = model.predict(X_test)

  accuracy = accuracy_score(y_test, pred_val)
  recall = recall_score(y_test, pred_val)
  precision = precision_score(y_test, pred_val)
  f1 = f1_score(y_test, pred_val)

  results.append({
      "Model": name,
      "Accuracy": accuracy,
      "Precision": precision,
      "Recall": recall,
      "F1 score": f1
  })

In [ ]:
results_df = pd.DataFrame(results)
results_df.set_index("Model", inplace=True)
results_df

,Accuracy,Precision,Recall,F1 score
Model,,,,
log_reg,0.750,0.558140,0.800000,0.657534
decision_tree,0.665,0.440678,0.433333,0.436975
random_forest,0.745,0.628571,0.366667,0.463158
svm_model,0.690,0.476190,0.333333,0.392157


### Insights

**Logistic Regression**: Recall increased substantially (0.512 -> 0.800), while accuracy and precision decreased.


**Decision Tree**: All evaluation metrics improved after applying class weights.


**Random Forest**: Performance remained almost unchanged, indicating minimal impact.


**SVM**: Recall and F1-score improved, but accuracy and precision decreased.

Overall, class weights improved the detection of the minority class for some models, but the improvement often came with a reduction in a precision or accuracy.


### 6.3 Experiment-3 (Hyperparameter Tuning)

In [ ]:
from sklearn.model_selection import GridSearchCV

# Define the parameter grid for logistic regression
# I want to test different levels of regualarisation,
# regularisation types, and class weighting
logistic_param_grid = {
    "C": [0.01, 0.1, 1, 10, 100],
    "penalty": ["l1", "l2"],
    "class_weight": [None, "balanced"]
}

# Define the parameter grid for random forest
# I want to test the number of trees and different levels
# of tree complexity
rf_param_grid = {
    "n_estimators": [500, 600, 900],
    "max_depth": [5, 10, 20, None],
    "min_samples_leaf": [1, 2, 5]
}

# Create the logistic Regression model
# GridSearchCV will test the parameter combinations above
log_reg = LogisticRegression(random_state=42, max_iter=10000)


# Set up GridSearchCV for Logistic Regression
# cv = 5 means each combination will be evaluated using 5-fold cross-validation.
# F1 score is used because both precision and recall matter for this project.
log_reg_grid = GridSearchCV(
    estimator = log_reg,
    param_grid = logistic_param_grid,
    cv=5,
    scoring="f1"
)


# Run the hyperparameter search on the training data
log_reg_grid.fit(X_train, y_train)

# Create the Random Forest model
rf = RandomForestClassifier(random_state=42)

# Set up GridSearchCV for Random Forest
rf_grid = GridSearchCV(
    estimator=rf,
    param_grid = rf_param_grid,
    cv=5,
    scoring="f1"
)

# Run the hyperparameter search on the training data
rf_grid.fit(X_train, y_train)


# Check the best parameter combination found for each model
print(log_reg_grid.best_params_)
print(log_reg_grid.best_score_)

print(rf_grid.best_params_)
print(rf_grid.best_score_)

/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

KeyboardInterrupt: 